assets/components/dialog/dialog.jsx

```diff
@@ -0,0 +1,115 @@
+// @flow
+
+// ----- Imports ----- //
+
+import React, { Component, type Node } from 'react';
+
+import { type Option } from 'helpers/types/option';
+import { classNameWithModifiers } from 'helpers/utilities';
+
+import './dialog.scss';
+
+
+// ----- Props ----- //
+
+export type PropTypes = {|
+  onStatusChange: (boolean) => void,
+  modal: boolean,
+  open: boolean,
+  'aria-label': Option<string>,
+  dismissOnBackgroundClick: boolean,
+  children: Node
+|};
+
+
+// ----- Component ----- //
+
+class Dialog extends Component<PropTypes> {
+
+  static defaultProps = {
+    onStatusChange: () => {},
+    modal: true,
+    open: false,
+    dismissOnBackgroundClick: false,
+  }
+
+  componentDidMount() {
+    if (this.props.open) {
+      this.open();
+    }
+  }
+
+  componentDidUpdate(prevProps: PropTypes) {
+    if (prevProps.open === true && this.props.open === false) {
+      this.close();
+    } else if (prevProps.open === false && this.props.open === true) {
+      this.open();
+    }
+  }
+
+  open() {
+    if (this.ref && this.ref.showModal) {
+      if (this.props.modal) {
+        this.ref.showModal();
+      } else {
+        this.ref.show();
+      }
+    }
+    requestAnimationFrame(() => {
+      if (this.ref) {
+        this.ref.focus();
+      }
+    });
+  }
+
+  close() {
+    if (this.ref && this.ref.close) {
+      this.ref.close();
+    }
+  }
+
+  ref: ?(HTMLDialogElement & {focus: Function});
+
+  render() {
+    const {
+      open, modal, children, onStatusChange, dismissOnBackgroundClick, ...otherProps
+    } = this.props;
+
+    return (
+      <dialog // eslint-disable-line jsx-a11y/no-redundant-roles
+        className={classNameWithModifiers('component-dialog', [modal ? 'modal' : null, open ? 'open' : null])}
+        aria-modal={modal}
+        aria-hidden={!open}
```

Comment:
Should we also include the `open` attribute?

Reply:
heya! it gets opened using the [js callbacks](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog) in the lifecycle hooks a bit higher up the component. You can't actually use both. if you have `open="true"` and try to fire up `dialog.show()` that's a javascript warning, makes for a lovely imperative vs declarative debate!

(Also, `open="true"` doesn't support modals at all 😝. I think the usage scenario is for serving markup from a server. Even the MDN demos don't set it via js)

As all `<dialog>` elements support `dialog.show()` I chose to omit the open attr altogether and instead style the polyfill with traditional css

Reply:
Interesting. And so confusing! I'm sure it's contrary to what is outlined in [the spec](https://html.spec.whatwg.org/multipage//interactive-elements.html#the-dialog-element). Browsers are a law unto themselves 🙄